README.md
# Touch Lua Emulator
**NOTE: This project is no longer in development. Although I can provide some support there will be no more releases.**

With this tool, you can run your Touch Lua programs on your computer, with little or no changes made to them. Right now it supports programs using draw and system library dating to Touch Lua 1.2 (before image and audio support were added) Lua SQLite3, Lua Matrix and Lua Complex are not included.

# Installation and use
- [Download](https://github.com/Croutonix/touch-lua-emulator/releases) the emulator and save it in any directory (you don't need to have lua installed)
- You can run example program by double-clicking on `run.bat`
- Open `config.lua` and adjust the settings (eg: screen size)
- To run your program, you can:
    - Drag and drop your file to `run.bat`
    - Edit `config.lua` and change file run by default and then run `run.bat`
    - Open command line in the emulator directory and do `run <filename>`
- Press ESC to exit, +/- to change zoom, 0 to reset zoom and ENTER to change screen orientation

## Screenshots
<img src="/example.png" width="300px" alt="Example program"/>

## Usage notes
- Your program might be accessing or writing files, and they are going to be written in the main directory. If you want them to be in another directory, you'll have to change the path in your code.
- If it has a main game loop which does not contain draw.doevents(), you won't be able to zoom, change orientation or close your program with ESC. (only by closing with console or killing the task) This is because draw.doevents() is used to do the emulator events too. If you don't want to do your program events, just the emulator's, add draw.doevents(false)
- I recommend adding a short delay (10ms) between each frame to prevent CPU overload.

## Fonts and alerts
Some fonts from your device are matched with equivalent fonts found on computers. There are only 6 matched fonts by default and only Helvetica supports text with styles (bold, italic...). If you need to add new fonts, add the font file in `res/font/` (needs to be .ttf) Then add a new match in config.lua: `["<device font>"] = "<filename.ttf>"`. If you need to add alerts, add the mp3 file in `res/sound/`. Could be useful to replace audio library for now?

# License
- Licensed under Apache 2.0
- Sounds are licensed under CC0 1.0 Universal
- I should not be using Helvetica but I am


CHANGELOG.md
# v0.1.1 (19/08/2016)
- Added support for sys.input(), draw.clearevents()
- Fixed console closing after program ends

# v0.1.0 (16/08/2016)
- Added support for draw.setscreen(), draw.settitle(), draw.clear(), draw.getsize(), draw.getport(), draw.string(), draw.stringinrect(), draw.point(), draw.moveto(), draw.lineto(), draw.line(), draw.rect(), draw.fillrect(), draw.circle(), draw.fillcircle(), draw.arc(), draw.fillarc(), draw.ellipse(), draw.fillellipse(), draw.polygon(), draw.fillpolygon(), draw.star(), draw.fillstar(), draw.triangle(), draw.filltriangle(), draw.roundedrect(), draw.fillroundedrect(), draw.stringsize(), draw.setfont(), draw.setlinestyle(), draw.setantialias(), draw.getantialias(), draw.sleep(), draw.gettime(), draw.refresh(), draw.beginframe(), draw.endframe(), draw.enablerefresh(), draw.disablerefresh(), draw.waittouch(), draw.tracktouches(), draw.doevents(), sys.print(), sys.println(), sys.clear(), sys.sleep(), sys.alert(), sys.gettime()
- Added console
- Added zoom and orientation

config.lua
local config = {
	file_name = "example.lua";  -- File to run on execution (leave blank for drag and drop)
	
	----- SCREEN SIZE OPTIONS -----
	portrait_size = {320, 460};   -- Screen viewable size on portrait orientation
	landscape_size = {480, 300};  -- Screen viewable size on landscape orientation
	
	----- DRAW & SCREEN OPTIONS (default values) -----
	antialias = true;          -- Default antialias value
	auto_refresh = true;       -- Default auto refresh value
	
	line_width = 1;            -- Default line width
	line_cap = "round";        -- Default line cap/join
	
	screen = 0;                -- Default screen on execution (0=console / 1=draw)
	orientation = "portrait";  -- Default screen orientation (portrait/landscape)
	zoom = 1;                  -- Default screen zoom (Between 0.5 and 2)
	window_title = "";         -- Default window title
	
	----- FONT OPTIONS -----
	font_name = "Helvetica";   -- Default font face
	font_size = 20;            -- Default font size
	
	-- Match fonts on your device with equivalent fonts on your computer
	-- Fonts must be placed in res/font folder
	-- Unmatched font will be set to default font
	fonts = {
		["Helvetica"]              = "segoeui.ttf";   -- Helvetica = Segoe UI
		["Helvetica-Bold"]         = "segoeuib.ttf";
		["Helvetica-Oblique"]      = "segoeuii.ttf";
		["Helvetica-Light"]        = "segoeuil.ttf";
		["Helvetica-BoldOblique"]  = "segoeuiz.ttf";
		["Helvetica-LightOblique"] = "seguili.ttf";
		["Times New Roman"]        = "times.ttf";     -- Times New Roman
		["Avenir"]                 = "calibri.ttf";   -- Avenir = Calibri
		["Chalkboard SE"]          = "comic.ttf";     -- Chalkboard SE = Comic Sans MS
		["Courier New"]            = "cour.ttf";      -- Courier New
		["Verdana"]                = "verdana.ttf";   -- Verdana
	}
}

return config

example.lua
-- Clock
function main()
   draw.setscreen(1)
   draw.settitle('Clock')

   while true do
      draw.doevents()
      draw.beginframe()
      draw.clear()
      getTime()
      drawClock()
      draw.endframe()
      draw.sleep(250)
   end
end

function drawClock()
   draw.setantialias(false)

   -- Draw the clock border
   draw.setlinestyle(2, 'round')
   draw.circle(px, py, radius, draw.black)

   -- Draw lines around the clock
   for i = 1, 60 do
      a = i*6 * math.pi/180
      b = 10 draw.setlinestyle(2, 'round')
      if i%5 == 0 then b = 15 draw.setlinestyle(3, 'round') end
      x1, y1 = px+(radius-1)*math.cos(a), py+(radius-1)*math.sin(a)
      x2, y2 = px+(radius-b)*math.cos(a), py+(radius-b)*math.sin(a)
      draw.line(x1, y1, x2, y2, draw.black)
   end

   -- Draw the hands
   angle, color = {}, {draw.darkgray, gray, red}
   angle[1] = (hour*6 + min/10 - 90) * math.pi / 180
   angle[2] = (min*6 + sec/10 - 90) * math.pi / 180
   angle[3] = (sec*6 - 90) * math.pi / 180
   for i = 1, 3 do
      draw.setlinestyle(7-i, 'round')
      draw.line(px, py, px+(radius-(5-i)*10)*math.cos(angle[i]), py+(radius-(5-i)*10)*math.sin(angle[i]), color[i])
   end

   -- Draw the clock's center
   draw.setlinestyle(10, 'round')
   draw.point(px, py, draw.black)

   -- Print time
   draw.setantialias(true)
   if string.len(sec) == 1 then sec = '0'..sec end
   if string.len(min) == 1 then min = '0'..min end
   if string.len(hour) == 1 then hour = '0'..hour end
   a = hour..':'..min..':'..sec
   draw.setfont('GillSans', 40)
   w = draw.stringsize(a)
   draw.string(a, px-w/2, py+radius+10, draw.black)

   -- Print date
   a = month..' '..day
   w, h = draw.stringsize(a)
   draw.string(a, px-w/2, py-radius-h-10, draw.black)
end

function getTime()
   time = os.date('*t')
   hour, min, sec = time.hour, time.min, time.sec
   month, day = months[time.month], time.day
   drawClock()
end

-- Months
months = {'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'}

-- Half transparent colors (hands)
red = {1, 0, 0, 0.5}
gray = {0.5, 0.5, 0.5, 0.5}

-- Size and position
px = draw.getport()
px, py = px/2, 250
radius = 120

main()

src/sys.lua
----------[[ SYSTEM LIBRARY ]]----------
local sys = {}

local alert_cache = {}
local alertdir = "res/sound/"

-- Print without adding new line
function sys.print(...)
	local arg = {...}
	local result = ""
  for i, v in ipairs(arg) do
    result = result .. tostring(v) 
		if i ~= #arg then result = result .. "\t" end
  end
	io.write(result)
end

-- Print by adding new line
function sys.println(...)
	sys.print(...)
	io.write("\n")
end

-- Input
function sys.input(input)
	io.write(input)
	return io.read()
end

-- Clear text screen
function sys.clear()
	os.execute("cls")
end

-- Wait a certain amount of time (ms)
function sys.sleep(values)
	love.timer.sleep(ms/1000)
end

-- Change cursor position
function sys.locate(row, column)
	-- Not implanted
end

-- Play alert sounds
function sys.alert(tone)
	local alert
	if alert_cache[tone] then
		alert = alert_cache[tone]
	else
		local file = io.open(alertdir..tone..".mp3", "rb")
		if not file then return end
		local content = file:read("*a")
		local data, err = love.filesystem.newFileData(content, tone..".mp3")
		local sound = love.sound.newSoundData(data)
		alert = love.audio.newSource(sound)
		alert_cache[tone] = alert
	end
	love.audio.play(alert)
end

-- Play alert sounds
function sys.gettime()
	return love.timer.getTime( )
end

-- Return table of filenames
function sys.dir(path)
	-- Not implanted
	return {}
end

return sys

src/draw.lua
----------[[ DRAW LIBRARY ]]----------
local draw = {}

----------- VARIABLES AND CONSTANTS -----------
-- Define colors
draw.white     = {1.00, 1.00, 1.00, 1.00}
draw.black     = {0.00, 0.00, 0.00, 1.00}
draw.red       = {1.00, 0.00, 0.00, 1.00}
draw.green     = {0.00, 1.00, 0.00, 1.00}
draw.blue      = {0.00, 0.00, 1.00, 1.00}
draw.cyan      = {0.00, 1.00, 1.00, 1.00}
draw.magenta   = {1.00, 0.00, 1.00, 1.00}
draw.orange    = {1.00, 0.50, 0.00, 1.00}
draw.purple    = {0.50, 0.00, 0.50, 1.00}
draw.yellow    = {1.00, 1.00, 0.00, 1.00}
draw.brown     = {0.60, 0.40, 0.20, 1.00}
draw.gray      = {0.50, 0.50, 0.50, 1.00}
draw.darkgray  = {0.33, 0.33, 0.33, 1.00}
draw.lightgray = {0.66, 0.66, 0.66, 1.00}

-- Screen
local screen_size, zoom, screen, canvas
local window_title = ""  -- Default window title

-- Refresh
local auto_refresh, old_auto_refresh, antialias

-- Events
local pressed_action, moved_action, released_action  -- Functions to call in case of an event
local ispressed = false  -- Is mouse pressed
local paused, resume_at = false, nil  -- Used for draw.waittouch()

-- Fonts
local fontdir = "res/font/"  -- Folder containing fonts
local font, default_font  -- Used to save current font
local fonts  -- Contains font matches
local font_cache = {}  -- Use to prevent rendering the same font again

-- Others
local lastX, lastY = {0}, {0}  -- Last lineto position
local line_width, line_join, line_style

----------- SCREEN FUNCTIONS -----------
-- Change between graphics and text mode
function draw.setscreen(n)
	if n == 0 then
		screen = 0
		love.window.close()
	elseif n == 1 then
		screen = 1
		setscreensize()  -- Set inital window size
		
		local s = screen_size
		canvas = love.graphics.newCanvas(math.max(s[2][1], s[2][2]), math.max(s[1][1], s[1][2]))  -- New canvas
		love.graphics.setCanvas(canvas)  -- Set render target to canvas
		draw.settitle(window_title)  -- Set title to current
		draw.clear()  -- Clear screen with white
		draw.refresh()
	end
end

-- Set graphics screen title
function draw.settitle(title)
	if screen == 0 then return end
	window_title = title
	love.window.setTitle(window_title)
end

-- Clear the screen
function draw.clear(color)
	if screen == 0 then return end
	if not color then color = draw.white end
	love.graphics.clear(convertcolor(color))
	if auto_refresh then draw.refresh() end
end

-- Return screen resolution
function draw.getsize()
    local _, _, flags = love.window.getMode()
    local w, h = love.window.getDesktopDimensions(flags.display)
	return w, h
end

-- Return viewable screen size
function draw.getport()
	return screen_size[orientation+1][1], screen_size[orientation+1][2]
end


----------- DRAW FUNCTIONS -----------
-- Draw string
function draw.string(text, x, y, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))
	love.graphics.print(text, x, y)
	if auto_refresh then draw.refresh() end
end

-- Draw string centered in rectangle
function draw.stringinrect(text, x1, y1, x2, y2, color)
	if screen == 0 then return end
	local w, h = draw.stringsize(text)
	local x, y = x1 + (x2-x1)/2 - w/2, y1 + (y2-y1)/2 - h/2
	love.graphics.setColor(convertcolor(color))
	love.graphics.print(text, x, y)
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw point
function draw.point(x, y, color)
	if screen == 0 then return end
	n = line_width/2
	if line_join == "round" then
		draw.fillcircle(x, y, n, color)
	else
		draw.fillrect(x-n, y-n, x+n, y+n, color)
	end
	if auto_refresh then draw.refresh() end
end

---------------------
-- Move current point to coordinate
function draw.moveto(x, y)
	lastX, lastY = {x}, {y}
end

-- Draw line from current point to coordinate
function draw.lineto(x, y, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))
	if lastX[2] then
		love.graphics.line(lastX[2],lastY[2],  lastX[1],lastY[1],  x,y)
	else
		love.graphics.line(lastX[1],lastY[1],  x,y)
	end
	if line_style == "round" then
		love.graphics.circle("fill", lastX[1], lastY[1], line_width/2)
		love.graphics.circle("fill", x, y, line_width/2)
	end
	lastX, lastY = {x, lastX[1]}, {y, lastY[1]}
	if auto_refresh then draw.refresh() end
end

-- Draw line
function draw.line(x1, y1, x2, y2, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.line(x1, y1, x2, y2)
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw rectangle
function draw.rect(x1, y1, x2, y2, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.rectangle("line", x1, y1, x2-x1, y2-y1)
	if auto_refresh then draw.refresh() end
end

-- Draw filled rectangle
function draw.fillrect(x1, y1, x2, y2, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.rectangle("fill", x1, y1, x2-x1, y2-y1)
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw circle
function draw.circle(x, y, r, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.circle("line", x, y, r)	
	if auto_refresh then draw.refresh() end
end

-- Draw filled circle
function draw.fillcircle(x, y, r, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.circle("fill", x, y, r)
	if antialias then
		local w = line_width
		love.graphics.setLineWidth(1)
		love.graphics.circle("line", x, y, r)
		love.graphics.setLineWidth(line_width)
	end
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw arc
function draw.arc(x, y, r, start_angle, end_angle, color)
	if screen == 0 then return end
	love.graphics.setLineJoin("miter")
	if start_angle > end_angle then end_angle = end_angle + 2*math.pi end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.arc("line", "pie", x, y, r, start_angle, end_angle)
	if auto_refresh then draw.refresh() end
	love.graphics.setLineJoin(line_join)
end

-- Draw filled arc
function draw.fillarc(x, y, r, start_angle, end_angle, color)
	if screen == 0 then return end
	if start_angle > end_angle then end_angle = end_angle + 2*math.pi end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.arc("fill", "pie", x, y, r, start_angle, end_angle)
	if antialias then
		local w = line_width
		love.graphics.setLineWidth(1)
		love.graphics.arc("line", "pie", x, y, r, start_angle, end_angle)
		love.graphics.setLineWidth(line_width)
	end
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw ellipse
function draw.ellipse(x1, y1, x2, y2, color)
	if screen == 0 then return end
	xr, yr = (x2-x1)/2, (y2-y1)/2
	love.graphics.setColor(convertcolor(color))	
	love.graphics.ellipse("line", x1+xr, y1+yr, xr, yr)	
	if auto_refresh then draw.refresh() end
end

-- Draw filled ellipse
function draw.fillellipse(x1, y1, x2, y2, color)
	if screen == 0 then return end
	xr, yr = (x2-x1)/2, (y2-y1)/2
	love.graphics.setColor(convertcolor(color))	
	love.graphics.ellipse("fill", x1+xr, y1+yr, xr, yr)
	if antialias then
		local w = line_width
		love.graphics.setLineWidth(1)
		love.graphics.ellipse("line", x1+xr, y1+yr, xr, yr)
		love.graphics.setLineWidth(line_width)
	end
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw polygon
function draw.polygon(x, y, r, sides, rotation, color)
	if screen == 0 then return end
	love.graphics.setLineJoin("miter")
	local points = {}
	for i = 1, sides do
		local angle = (i/sides)*(2*math.pi) - rotation
		local px = math.cos(angle)*r + x
		local py = math.sin(angle)*r + y
		points[i*2-1], points[i*2] = px, py
	end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.polygon("line", points)
	if auto_refresh then draw.refresh() end
	love.graphics.setLineJoin(line_join)	
end

-- Draw filled polygon
function draw.fillpolygon(x, y, r, sides, rotation, color)
	if screen == 0 then return end
	local points = {}
	for i = 1, sides do
		local angle = (i/sides)*(2*math.pi) - rotation
		local px = math.cos(angle)*r + x
		local py = math.sin(angle)*r + y
		points[i*2-1], points[i*2] = px, py
	end
	love.graphics.setColor(convertcolor(color))	
	love.graphics.polygon("fill", points)
	if antialias then
		local w = line_width
		love.graphics.setLineWidth(1)
		love.graphics.polygon("line", points)
		love.graphics.setLineWidth(line_width)
	end
	if auto_refresh then draw.refresh() end	
end

---------------------
-- Draw star
function draw.star(x, y, inr, outr, vertices, rotation, color)
	if screen == 0 then return end
	love.graphics.setLineJoin("miter")

  local pts = {}	
  local rot = -rotation
  local step = math.pi/vertices
  for i = 1, vertices do
    pts[#pts+1] = x + math.cos(rot)*outr
    pts[#pts+1] = y + math.sin(rot)*outr
    rot = rot + step

    pts[#pts+1] = x + math.cos(rot)*inr
    pts[#pts+1] = y + math.sin(rot)*inr
    rot = rot + step
  end
	
	love.graphics.setColor(convertcolor(color))
	love.graphics.polygon("line", pts)
	if auto_refresh then draw.refresh() end
	love.graphics.setLineJoin(line_join)
end

-- Draw filled star
function draw.fillstar(x, y, inr, outr, vertices, rotation, color)
	if screen == 0 then return end
	love.graphics.setLineJoin("miter")

  local pts = {}	
  local rot = -rotation
  local step = math.pi/vertices
  for i = 1, vertices do
    pts[#pts+1] = x + math.cos(rot)*outr
    pts[#pts+1] = y + math.sin(rot)*outr
    rot = rot + step

    pts[#pts+1] = x + math.cos(rot)*inr
    pts[#pts+1] = y + math.sin(rot)*inr
    rot = rot + step
  end
	
	love.graphics.setColor(convertcolor(color))
	local triangles = love.math.triangulate(pts)
	for i, t in ipairs(triangles)	do
		love.graphics.polygon("fill", t[1], t[2], t[3], t[4], t[5], t[6])
	end
	if antialias then
		local w = line_width
		love.graphics.setLineWidth(1)
		love.graphics.polygon("line", pts)
		love.graphics.setLineWidth(line_width)
	end
	
	if auto_refresh then draw.refresh() end
	love.graphics.setLineJoin(line_join)
end

---------------------
-- Draw triangle
function draw.triangle(x1, y1, x2, y2, x3, y3, color)
	if screen == 0 then return end
	love.graphics.setLineJoin("miter")
	love.graphics.setColor(convertcolor(color))
	love.graphics.polygon("line", x1, y1, x2, y2, x3, y3)
	if auto_refresh then draw.refresh() end
	love.graphics.setLineJoin(line_join)
end

-- Draw filled triangle
function draw.filltriangle(x1, y1, x2, y2, x3, y3, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))
	love.graphics.polygon("fill", x1, y1, x2, y2, x3, y3)
		if antialias then love.graphics.polygon("line", x1, y1, x2, y2, x3, y3) end
	if auto_refresh then draw.refresh() end
end

---------------------
-- Draw rounded rectangle
function draw.roundedrect(x1, y1, x2, y2, r, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))
	love.graphics.arc("line", "open", x2-r, y2-r, r, 0,           math.pi*0.5)  -- Bot right
	love.graphics.arc("line", "open", x1+r, y2-r, r, math.pi*0.5, math.pi*1.0)  -- Bot left
	love.graphics.arc("line", "open", x1+r, y1+r, r, math.pi*1.0, math.pi*1.5)  -- Top right
	love.graphics.arc("line", "open", x2-r, y1+r, r, math.pi*1.5, math.pi*2.0)  -- Top left
	love.graphics.line(x1+r, y1, x2-r, y1)  -- Top
	love.graphics.line(x1+r, y2, x2-r, y2)  -- Bottom
	love.graphics.line(x1, y1+r, x1, y2-r)  -- Left
	love.graphics.line(x2, y1+r, x2, y2-r)  -- Right
	if auto_refresh then draw.refresh() end
end

-- Draw filled rounded rectangle
function draw.fillroundedrect(x1, y1, x2, y2, r, color)
	if screen == 0 then return end
	love.graphics.setColor(convertcolor(color))
	love.graphics.arc("fill", "pie", x2-r, y2-r, r, 0,           math.pi*0.5)  -- Bot right
	love.graphics.arc("fill", "pie", x1+r, y2-r, r, math.pi*0.5, math.pi*1.0)  -- Bot left
	love.graphics.arc("fill", "pie", x1+r, y1+r, r, math.pi*1.0, math.pi*1.5)  -- Top right
	love.graphics.arc("fill", "pie", x2-r, y1+r, r, math.pi*1.5, math.pi*2.0)  -- Top left
	love.graphics.rectangle("fill", x1+r, y1, x2-x1-2*r, r)            -- Top
	love.graphics.rectangle("fill", x1+r, y2-r, x2-x1-2*r, r)          -- Bottom
	love.graphics.rectangle("fill", x1, y1+r, r, y2-y1-2*r)            -- Left
	love.graphics.rectangle("fill", x2-r, y1+r, r, y2-y1-2*r)          -- Right
	love.graphics.rectangle("fill", x1+r, y1+r, x2-x1-2*r, y2-y1-2*r)  -- Center
	if auto_refresh then draw.refresh() end
end


----------- DRAW PARAMETERS AND OTHER -----------
-- Return drawn string size
function draw.stringsize(text)
	local w = font:getWidth(text)
	local h = font:getHeight()
	return w, h
end

-- Set font face and size
function draw.setfont(name, size)
	name = fonts[name]
	if name == nil then name = fonts[default_font] end
	
	if font_cache[name] and font_cache[name][size] then
		font = font_cache[name][size]
	else
		local file = io.open(fontdir..name, "rb")
		local content = file:read("*a")
		local data, err = love.filesystem.newFileData(content, name)
		font = love.graphics.newFont(data, size)
		
		if not font_cache[name] then font_cache[name] = {} end
		font_cache[name][size] = font
	end
	love.graphics.setFont(font)
end

-- Set line width and ending
function draw.setlinestyle(width, style)
	line_width, line_style = width, style
	love.graphics.setLineWidth(line_width)  -- Set line width
	 
	if style == "square" then  -- Set line join/style
		love.graphics.setLineJoin("miter")  -- Square
		line_join = "miter"
	else  
		love.graphics.setLineJoin("none")   -- Butt/round
		line_join = "none"
	end
end

-- Toggle antialias
function draw.setantialias(flag)
	antialiasing = flag
	if antialiasing ~= true and antialiasing ~= false then antialiasing = true end
	if antialiasing then
		love.graphics.setLineStyle("smooth")  
	else
		love.graphics.setLineStyle("rough")
	end
end

-- Get antialias state
function draw.getantialias()
	return antialiasing
end


----------- TIME FUNCTIONS -----------
-- Wait a certain amount of time
function draw.sleep(ms)
	love.timer.sleep(ms/1000)
end

-- Return system time in seconds
function draw.gettime()
	return love.timer.getTime( )
end


----------- DISPLAY FUNCTIONS -----------
-- Display current frame 
function draw.refresh()
	if screen == 0 then return end
	love.graphics.setColor(255, 255, 255, 255)
	love.graphics.origin()
	love.graphics.setCanvas()       -- Set target to screen
	love.graphics.draw(canvas, 0, 0, 0, zoom)  -- Draw canvas to screen
	love.graphics.present()         -- Refresh screen
	love.graphics.setCanvas(canvas) -- Set target to canvas
end

-- Disable auto refresh until draw.endframe()
function draw.beginframe()
	old_auto_refresh = auto_refresh
	auto_refresh = false
end

-- Display frame and reenable auto refresh
function draw.endframe()
	auto_refresh = old_auto_refresh
	draw.refresh()
end

-- Enable auto screen refresh
function draw.enablerefresh()
	auto_refresh = true
end

-- Disable auto screen refresh
function draw.disablerefresh()
	auto_refresh = false
end


----------- EVEMT FUNCTIONS -----------
-- Wait for user to make touch
function draw.waittouch()
	if screen == 0 then return end
	paused = true
	while paused do
		draw.doevents()
	end
	return resume_at[1], resume_at[2]
end

-- Touch handling
function draw.tracktouches(p, m, r)
	if screen == 0 then return end
	pressed_action, moved_action, released_action = p, m, r  -- Save function
end

-- Call function related to touches
function draw.doevents(bool)
	if screen == 0 then return end
	if bool == nil then bool = true end  -- if bool == false then doevents is used for exit
	
	if love.event then
		love.event.pump()
		for event, a,b,c,d,e,f in love.event.poll() do
			if bool then 
				if event == "mousepressed" then  -- Mouse pressed
					ispressed = true
					local x, y = math.floor((a/zoom)+0.5), math.floor((b/zoom)+0.5)
					if pressed_action then pressed_action(x, y) end
					
					if paused then
						paused = false
						resume_at = {x, y}
					end
				end
				
				if ispressed and event == "mousemoved" then  -- Mouse moved while pressed
					local x, y = math.floor((a/zoom)+0.5), math.floor((b/zoom)+0.5)
					if moved_action then moved_action(x, y) end
				end
			
				if event == "mousereleased" then  -- Mouse released
					ispressed = false
					local x, y = math.floor((a/zoom)+0.5), math.floor((b/zoom)+0.5)
					if released_action then released_action(x, y) end
				end
			end
			
			if event == "keypressed" then keypressed(a, b, c) end  -- Key pressed
		end
	end
end

-- Reset events
function draw.clearevents()
	pressed_action, moved_action, released_action = nil, nil, nil
end

----------- EMULATOR-SPECIFIC FUNCTIONS -----------
-- Set appropriate window size
function setscreensize()
	if screen ~= 0 then
		local w, h = draw.getport()
		w, h = w*zoom, h*zoom  -- Apply zoom to window size
		love.graphics.setCanvas()
		love.window.setMode(w, h, {resizable=false, centered=true, vsync=true})  -- Set window size and position
		love.graphics.setCanvas(canvas)
	end
end

-- Key is pressed
function keypressed(key, scancode, isrepeat)
	if not isrepeat then
		-- Zoom in and out
		local oldzoom = zoom
		if key == "=" and zoom < 2 then zoom = zoom + 0.1 end    -- Zoom in
		if key == "-" and zoom > 0.5 then zoom = zoom - 0.1 end  -- Zoom out
		if key == "0" then zoom = 1 end                          -- Default zoom
		if zoom ~= oldzoom then apply_zoom(zoom) end
		
		-- Swap between orientations
		if key == "return" then apply_orientation(1 - orientation) end  

		-- Quit
		if key == "escape" then
			draw.setscreen(0)
			print("Program Break:")
			print("Interrupted by user")
			print() os.execute("pause")  -- Press any key to continue
			os.exit()
		end
	end
end

-- Set screen zoom
function apply_zoom(n)
	zoom = n
	setscreensize() -- Apply new window size
	draw.refresh()
end

-- Set screen orientation
function apply_orientation(n)
	orientation = n
	setscreensize()  -- Apply new orientation
	draw.refresh()
end

-- Convert color from 0-100% to 0-255
function convertcolor(c)
	return c[1]*255, c[2]*255, c[3]*255, c[4]*255
end

-- Apply settings
function apply_settings(config)
	screen_size = {config.portrait_size, config.landscape_size}
	antialias, auto_refresh = config.antialias, config.auto_refresh
	
	line_style = config.line_cap
	draw.setlinestyle(config.line_width, config.line_cap)
	
	fonts = config.fonts
	draw.setfont(config.font_name, config.font_size)
	default_font = config.font_name
	
	draw.setscreen(config.screen)
	draw.settitle(config.window_title)
	
	zoom, orientation = config.zoom, config.orientation
	if orientation == "portrait" then orientation = 0 end
	if orientation == "landscape" then orientation = 1 end
	apply_orientation(orientation)
	apply_zoom(zoom)
end

return draw